File: AFD.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

/* funcion que verifica si un caracter es mayuscula
   parametros:
   c: caracter a verificar
   retorna: 1 si es mayuscula A-Z, 0 en caso contrario */
int es_mayuscula(char c) {
    return c >= 'A' && c <= 'Z';
}

/* funcion que verifica si un caracter es minuscula o digito
   parametros:
   c: caracter a verificar
   retorna: 1 si es minuscula a-z o digito 0-9, 0 en caso contrario */
int es_minuscula_o_digito(char c) {
    return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
}

/* procesa una linea de texto

   parametros:
   linea: puntero a cadena con la linea a procesar

   variables internas:
   i: indice entero que recorre la cadena
   n: longitud de la cadena

   comportamiento:
   imprime tokens reconocidos por linea: INCR, SUMA, ID
   si encuentra un caracter no valido imprime NO ACEPTA y retorna 0
   retorna 1 en caso de exito
*/
int analizar_linea(char *linea) {
    int i = 0; /* indice actual */
    int n = strlen(linea); /* longitud de la linea */

    while (i < n) {
        /* ignorar espacios, saltos de linea y tabs */
        if (isspace(linea[i])) {
            i++;
            continue;
        }

        /* caso para suma (+) e incremento (++) */
        if (linea[i] == '+') {
            if (i + 1 < n && linea[i+1] == '+') {
                printf("INCR\n"); /* token incremento */
                i += 2; /* consumir los dos signos */
            } else {
                printf("SUMA\n"); /* token suma simple */
                i++; /* consumir un signo */
            }
            continue;
        }

        /* caso para identificador: empieza con mayuscula seguido
           de minusculas o digitos [A-Z][a-z0-9]* */
        if (es_mayuscula(linea[i])) {
            i++; /* consumir la letra inicial mayuscula */
            /* consumir todos los caracteres validos siguientes */
            while (i < n && es_minuscula_o_digito(linea[i])) {
                i++;
            }
            printf("ID\n"); /* token identificador */
            continue;
        }

        /* si no coincide con ninguna regla, no acepta */
        printf("NO ACEPTA\n");
        return 0; /* retorna error para detener el proceso */
    }
    return 1; /* exito */
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("uso: %s entrada.txt\n", argv[0]);
        return 1;
    }

    FILE *archivo = fopen(argv[1], "r");
    if (archivo == NULL) {
        printf("error: el archivo no existe\n");
        return 1;
    }

    char buffer[1024]; /* buffer para leer cada linea */
    /* leer el archivo linea por linea */
    while (fgets(buffer, sizeof(buffer), archivo)) {
        /* quitar el salto de linea al final si existe */
        buffer[strcspn(buffer, "\n")] = '\0';

        /* si la linea esta vacia saltarla */
        if (strlen(buffer) == 0) {
            continue;
        }
        /* procesar la linea */
        analizar_linea(buffer);
    }

    fclose(archivo); /* cerrar el archivo */
    return 0;
}

File: AFD.py
import sys


def analizar_lexico(texto):
    """funcion que realiza el analisis lexico del texto.

    parametros:
    texto: cadena de entrada que sera analizada caracter por caracter

    variables internas:
    i: indice entero que recorre la cadena
    n: longitud de la cadena
    tokens_encontrados: lista donde se almacenan los tokens detectados

    retorna:
    lista de tokens o None si se encuentra un caracter no valido
    """
    i = 0 
    n = len(texto)  
    tokens_encontrados = []  

    while i < n:  # recorrer mientras queden caracteres
        char = texto[i]  # caracter actual en la posicion i

        # si el caracter es espacio en blanco, avanzar
        if char.isspace():
            i += 1
            continue

        # detectar suma (+) o incremento (++)
        if char == '+':
            # si el siguiente caracter tambien es '+', es incremento
            if i + 1 < n and texto[i+1] == '+':
                tokens_encontrados.append("INCR")  # token de incremento
                i += 2  # saltar los dos signos ++
            else:
                tokens_encontrados.append("SUMA")  # token de suma simple
                i += 1  # avanzar un caracter
            continue

        # detectar identificadores que empiezan con mayuscula A-Z
        if 'A' <= char <= 'Z':
            i += 1  # consumir la letra inicial
            # mientras sigan minusculas o numeros, continuar consumiendo
            while i < n and (('a' <= texto[i] <= 'z') or ('0' <= texto[i] <= '9')):
                i += 1
            tokens_encontrados.append("ID")  # token identificador
            continue

        # si se alcanza un caracter no esperado, devolver error (None)
        return None

    return tokens_encontrados  # devolver la lista de tokens reconocidos


def main():
    if len(sys.argv) != 2:
        return

    try:
        with open(sys.argv[1], 'r') as archivo:
            for linea in archivo:
                linea = linea.strip()
                if not linea:
                    continue
                
                resultado = analizar_lexico(linea)
                
                if resultado is None:
                    print("NO ACEPTA")
                else:
                    for t in resultado:
                        print(t)
    except FileNotFoundError:
        print("error: archivo no encontrado")

if __name__ == "__main__":
    main()